portlist/portlist_linux.go

```diff
@@ -25,96 +25,133 @@ import (
 	"tailscale.com/util/mak"
 )
 
-// Reading the sockfiles on Linux is very fast, so we can do it often.
-const pollInterval = 1 * time.Second
-
-var sockfiles = []string{"/proc/net/tcp", "/proc/net/tcp6", "/proc/net/udp", "/proc/net/udp6"}
-
-var sawProcNetPermissionErr atomic.Bool
-
-const (
-	v6Localhost = "00000000000000000000000001000000:"
-	v6Any       = "00000000000000000000000000000000:0000"
-	v4Localhost = "0100007F:"
-	v4Any       = "00000000:0000"
-)
+func init() {
+	newOSImpl = newLinuxImpl
+}
 
-var eofReader = bytes.NewReader(nil)
+type linuxImpl struct {
+	procNetFiles []*os.File // seeked to start & reused between calls
 
-var bufioReaderPool = &sync.Pool{
-	New: func() any { return bufio.NewReader(eofReader) },
+	known map[string]*portMeta // inode string => metadata
+	br    *bufio.Reader
 }
 
-type internedStrings struct {
-	m map[string]string
+type portMeta struct {
+	port          Port
+	keep          bool
+	needsProcName bool
 }
 
-func (v *internedStrings) get(b []byte) string {
-	if s, ok := v.m[string(b)]; ok {
-		return s
+func newLinuxImplBase() *linuxImpl {
+	return &linuxImpl{
+		br:    bufio.NewReader(eofReader),
+		known: map[string]*portMeta{},
 	}
-	s := string(b)
-	mak.Set(&v.m, s, s)
-	return s
 }
 
-var internedStringsPool = &sync.Pool{
-	New: func() any { return new(internedStrings) },
+func newLinuxImpl() osImpl {
+	li := newLinuxImplBase()
+	for _, name := range []string{
+		"/proc/net/tcp",
+		"/proc/net/tcp6",
+		"/proc/net/udp",
+		"/proc/net/udp6",
+	} {
+		f, err := os.Open(name)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			log.Printf("portlist warning; ignoring: %v", err)
+			continue
+		}
+		li.procNetFiles = append(li.procNetFiles, f)
+	}
+	return li
 }
 
-func appendListeningPorts(base []Port) ([]Port, error) {
-	ret := base
-	if sawProcNetPermissionErr.Load() {
-		return ret, nil
+func (li *linuxImpl) Close() error {
+	for _, f := range li.procNetFiles {
+		f.Close()
 	}
+	li.procNetFiles = nil
+	return nil
+}
 
-	br := bufioReaderPool.Get().(*bufio.Reader)
-	defer bufioReaderPool.Put(br)
-	defer br.Reset(eofReader)
+// Reading the sockfiles on Linux is very fast, so we can do it often.
+const pollInterval = 1 * time.Second
 
-	stringCache := internedStringsPool.Get().(*internedStrings)
-	defer internedStringsPool.Put(stringCache)
+const (
+	v6Localhost = "00000000000000000000000001000000:"
+	v6Any       = "00000000000000000000000000000000:0000"
+	v4Localhost = "0100007F:"
+	v4Any       = "00000000:0000"
+)
+
+var eofReader = bytes.NewReader(nil)
 
-	for _, fname := range sockfiles {
+func (li *linuxImpl) AppendListeningPorts(base []Port) ([]Port, error) {
+	if runtime.GOOS == "android" {
```

Comment:
In the older code, Android 9 and earlier use this Linux implementation. I don't actually know what Android 10 and later have been doing here, do we know what the older Android releases will do when they run that Android code?

If not, we can figure it out the next time we update OSS in tailscale-android.

Reply:
I'm basically giving up on all this code on Android. It's not worth maintaining for limited benefit (servers are rare on Android) and decreasing support (Android 10+ is only becoming more common over time.... in fact there are barely any Android 9 and earlier these days percentage-wise.